### PR TITLE
Save & share project

### PR DESCRIPTION
**Pull request by PierreCrb**
Added functionality to be able to save / share an OpenChakra project from the URL ✨

- Creation of the project if no id is specified in the url ✅

- Update of the already existing project if the id is present in the url and the user id is the correct one ✅

- Be able to share your project with the following url: https://openchakra.app/project/:id ✅

Bug 🚫

When a project has just been created, the user is redirected to his project page but this error appears:

`Cannot have two HTML5 backends at the same time.`

When we refresh the page, the bug disappears, I have not managed to fix this bug for the moment 💡

### CODE REVIEW

**Comment by PierreCrb** [comment]
`Cannot have two HTML5 backends at the same time.`

Bug fix ! you can save your project ✅

File: src/components/Header.tsx [hunk 1]
```diff
@@ -65,7 +65,7 @@ const CodeSandboxButton = () => {
   )
 }
 
-const Header = () => {
+const Header = ({ id }: any) => {
```
Comment (shinework) [inline src/components/Header.tsx:68]: Can you replace `any` by the right types?

File: src/components/headerMenu/HeaderMenu.tsx [hunk 2]
```diff
@@ -40,7 +40,7 @@ const ExportMenuItem = dynamic(() => import('./ExportMenuItem'), { ssr: false })
 const ImportMenuItem = dynamic(() => import('./ImportMenuItem'), { ssr: false })
 const SaveMenuItem = dynamic(() => import('./SaveMenuItem'), { ssr: false })
 
-const HeaderMenu = () => {
+const HeaderMenu = (props: any) => {
```
Comment (shinework) [inline src/components/headerMenu/HeaderMenu.tsx:43]: Can you replace `any` by the right types?

File: src/components/headerMenu/SaveMenuItem.tsx [hunk 3]
```diff
@@ -2,14 +2,71 @@ import React from 'react'
...
 import { FaSave } from 'react-icons/fa'
-import { saveAsJSON } from '~utils/import'
 import { getComponents } from '~core/selectors/components'
+import { signIn, useSession } from 'next-auth/client'
+import { useRouter } from 'next/router'
+import useDispatch from '~hooks/useDispatch'
+import { checkUser, createProject } from '~utils/checkProject'
 
-const SaveMenuItem = () => {
+const SaveMenuItem = (props: any) => {
```
Comment (shinework) [inline src/components/headerMenu/SaveMenuItem.tsx:11]: Can you replace `any` by the right types?

File: src/components/headerMenu/SaveMenuItem.tsx [hunk 4]
```diff
@@ -2,14 +2,71 @@ import React from 'react'
...
+
+  const updateProject = async () => {
+    const markup = JSON.stringify(components)
+    let bodyData = {
+      project: {
+        markup: markup,
+        id: props.id,
+      },
+    }
+    const response = await fetch('http://localhost:3000/api/project/update', {
```
Comment (shinework) [inline src/components/headerMenu/SaveMenuItem.tsx:25]:
```suggestion
    const response = await fetch('/api/project/update', {
```

File: src/components/headerMenu/SaveMenuItem.tsx [hunk 5]
```diff
@@ -2,14 +2,71 @@ import React from 'react'
...
+    const { project } = data
+    return project
+  }
+
+  const saveProject = async () => {
+    if (session) {
+      if (props.id) {
+        let userCanEdit = false
+        const userProject = await checkUser(session.user.name)
+        userProject.project.map((e: any) => {
```
Comment (shinework) [inline src/components/headerMenu/SaveMenuItem.tsx:42]: Do not use `any` 😉

File: src/pages/api/project/check.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import { getSession } from 'next-auth/client'
+
+export default async function(req: NextApiRequest, res: NextApiResponse) {
+  const prisma = new PrismaClient({ log: ['query'] })
+  const session = await getSession({ req })
+
+  try {
+    const actualUser = await prisma.user.findMany({
+      where: {
+        name: session!.user.name,
```
Comment (shinework) [inline src/pages/api/project/check.tsx:12]: Maybe it's safer to fetch the user by his id?

File: src/pages/api/project/create.tsx [hunk 7]
```diff
@@ -1,22 +1,30 @@
...
 
 export default async function(req: NextApiRequest, res: NextApiResponse) {
   const prisma = new PrismaClient({ log: ['query'] })
-
+  const session = await getSession({ req })
   try {
     const { project: projectData } = req.body
+    const actualUser = await prisma.user.findMany({
+      where: {
+        name: session!.user.name,
```
Comment (shinework) [inline src/pages/api/project/create.tsx:12]: Same here (id)

File: src/pages/api/project/index.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { PrismaClient } from '@prisma/client'
+
+export default async function(req: NextApiRequest, res: NextApiResponse) {
+  const prisma = new PrismaClient({ log: ['query'] })
```
Comment (shinework) [inline src/pages/api/project/index.tsx:5]: remove log

File: src/pages/api/project/update.tsx [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { PrismaClient } from '@prisma/client'
+
+export default async function(req: NextApiRequest, res: NextApiResponse) {
+  const prisma = new PrismaClient({ log: ['query'] })
```
Comment (shinework) [inline src/pages/api/project/update.tsx:5]: remove log

File: src/pages/api/project/index.tsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+  try {
+    const projects = await prisma.project.findMany()
+    res.status(200)
+    res.json({ projects })
+  } catch (e) {
+    res.status(500)
+    res.json({ error: 'Unable to fetch project' })
+  } finally {
+    await prisma.$disconnect
```
Comment (shinework) [inline src/pages/api/project/index.tsx:15]: 👍

File: src/pages/api/user/index.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { PrismaClient } from '@prisma/client'
+
+export default async function(req: NextApiRequest, res: NextApiResponse) {
+  const prisma = new PrismaClient({ log: ['query'] })
```
Comment (shinework) [inline src/pages/api/user/index.tsx:5]: remove :)

File: src/pages/index.tsx [hunk 12]
```diff
@@ -13,9 +13,8 @@ import Editor from '~components/editor/Editor'
 import { InspectorProvider } from '~contexts/inspector-context'
 import Inspector from '~components/inspector/Inspector'
 
-const App = () => {
+const App = ({ id }: any) => {
```
Comment (shinework) [inline src/pages/index.tsx:16]: Don't use any :)

File: src/pages/index.tsx [hunk 13]
```diff
@@ -26,7 +25,8 @@ const App = () => {
 
       <Metadata />
 
-      <Header />
+      <Header id={id} />
```
Comment (shinework) [inline src/pages/index.tsx:28]: The `id` may not exist no?

File: src/utils/checkProject.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
+export const checkUser = async (name: string) => {
+  const response = await fetch('http://localhost:3000/api/project/check', {
```
Comment (shinework) [inline src/utils/checkProject.ts:2]:
```suggestion
  const response = await fetch('/api/project/check', {
```

File: src/utils/checkProject.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  return data
+}
+
+export const createProject = async (markup: string) => {
+  let bodyData = {
+    project: {
+      markup: markup,
+    },
+  }
+  const response = await fetch('http://localhost:3000/api/project/create', {
```
Comment (shinework) [inline src/utils/checkProject.ts:19]:
```suggestion
  const response = await fetch('/api/project/create', {
```

**Review by shinework — CHANGES_REQUESTED** [review]

**Comment by PierreCrb** [comment]
Adding a spinner for loading if an id is specified in the url to then display the project

loading

**Review by shinework — APPROVED** [review]

**Pull request merged**